README.md
# SSHKit

[![Build Status](https://travis-ci.org/bitcrowd/sshkit.ex.svg?branch=master)](https://travis-ci.org/bitcrowd/sshkit.ex)
[![Inline docs](https://inch-ci.org/github/bitcrowd/sshkit.ex.svg?branch=master)](https://inch-ci.org/github/bitcrowd/sshkit.ex)

SSHKit is an Elixir toolkit for performing tasks on one or more servers, built on top of Erlang’s SSH application.

[Documentation for SSHKit is available online][docs].

## Usage

SSHKit is designed to enable server task automation in a structured and repeatable way, e.g. in the context of deployment tools:

```elixir
hosts = ["1.eg.io", {"2.eg.io", port: 2222}]

{:ok, conn} = SSHKit.connect(hosts)

{:ok, _} = SSHKit.run(conn, "apt-get update -y")

{:ok, } = SSHKit.stream(chan)

context =
  SSHKit.context()
  |> SSHKit.path("/var/www/phx")
  |> SSHKit.user("deploy")
  |> SSHKit.group("deploy")
  |> SSHKit.umask("022")
  |> SSHKit.env(%{"NODE_ENV" => "production"})

{:ok, _} = SSHKit.upload(conn, ".", recursive: true, context: context)

# TODO: Receive upload status messages

{:ok, chan} = SSHKit.run(conn, "yarn install", context: context)

# TODO: Showcase streaming interface

:ok = SSHKit.close(conn)
```

The [`SSHKit`](https://hexdocs.pm/sshkit/SSHKit.html) module documentation has more guidance and examples for the DSL.

## Installation

Just add `sshkit` to your list of dependencies in `mix.exs`:

  ```elixir
  def deps do
    [{:sshkit, "~> 1.0"}]
  end
  ```

SSHKit should be automatically started unless the `:applications` key is set inside `def application` in your `mix.exs`. In such cases, you need to [remove the `:applications` key in favor of `:extra_applications`](https://elixir-lang.org/blog/2017/01/05/elixir-v1-4-0-released/#application-inference).

## Testing

As usual, to run all tests, use:

```shell
mix test
```

Apart from unit tests, we also have [functional tests](https://en.wikipedia.org/wiki/Functional_testing). These check SSHKit against real SSH server implementations running inside Docker containers. Therefore, you need to have [Docker](https://www.docker.com/) installed.

All functional tests are tagged as such. Hence, if you wish to skip them:

```shell
mix test --exclude functional
```

Hint: We've found functional tests to run significantly faster with [Docker Machine](https://docs.docker.com/machine/) compared to [Docker for Mac](https://docs.docker.com/docker-for-mac/) on OS X.

## Releasing

* Make sure tests pass: `mix test`.
* Increase version number in `mix.exs`, keeping [semantic versioning](https://semver.org/) in mind.
* Update [CHANGLOG.md][changelog]:
  * Create a new section for the current version.
  * Reset the `master` section to the empty template.
* Commit your changes: `git commit -m "Release 0.1.0"`.
* Tag the commit with the version number: `git tag -a v0.1.0`.
  * Annotate the tag with the respective section from [CHANGLOG.md][changelog] *(in a git-compatible format)*.
* Push your commit: `git push`.
* Push the tag: `git push origin v0.1.0`
* Publish the new release to hex.pm: `mix hex.publish`.
  * You can find the hex.pm credentials in the bitcrowd password store.

## Contributing

We welcome everyone to contribute to SSHKit and help us tackle existing issues!

Use the [issue tracker][issues] for bug reports or feature requests. Open a [pull request][pulls] when you are ready to contribute.

If you are planning to contribute documentation, please check [the best practices for writing documentation][writing-docs].

## Thanks

SSHKit is inspired by [SSHKit for Ruby](https://github.com/capistrano/sshkit) which is part of the fantastic [Capistrano](https://github.com/capistrano) project.

It deliberately departs from its role model with regard to its API given the very different nature of the two programming languages.

If you are looking for an Elixir deployment tool similar to Capistrano, take a look at [Bootleg](https://github.com/labzero/bootleg) which is based on top of SSHKit.

## License

SSHKit source code is released under the MIT License.

Check the [LICENSE][license] file for more information.

[issues]: https://github.com/bitcrowd/sshkit.ex/issues
[pulls]: https://github.com/bitcrowd/sshkit.ex/pulls
[docs]: https://hexdocs.pm/sshkit
[changelog]: ./CHANGELOG.md
[license]: ./LICENSE
[writing-docs]: https://hexdocs.pm/elixir/writing-documentation.html


CHANGELOG.md
# SSHKit Changelog

Presented in reverse chronological order.

## master

https://github.com/bitcrowd/sshkit.ex/compare/v0.2.0...HEAD

<!-- Put high-level summary here -->

<!-- Give thanks to contributors and mention them! -->

### Deprecations:

<!-- Put deprecations here -->

### Potentially breaking changes:

<!-- Put potentially breaking changes here -->

### New features:

<!-- Put new features here -->
* Add support to enable the shell on the server end using `SSH.Channel.shell`.

### Fixes:

<!-- Put fixes here -->

## `0.3.0` (2019-10-28)

https://github.com/bitcrowd/sshkit.ex/compare/v0.2.0...v0.3.0

This version adds SSH subsystem support.

Shout-out to @bartj3 🙌

### New features:

* Add support for initializing subsystems using `SSH.Channel.subsystem`
  * Subsystems allow applications or functions to use SSH as a protocol.
    Examples of this are SFTP and NETCONF over SSH.

## `0.2.0` (2019-10-17)

https://github.com/bitcrowd/sshkit.ex/compare/v0.1.0...v0.2.0

* The `SSHKit.download/3` and `SSHKit.upload/3` functions are now fully context-aware [#121]
* They will respect the `user`, `env`, `path`… values set in the context

### Fixes:

* Improve documentation about our release process
* Fix an Elixir 1.8 deprecation warning (#147)

## `0.1.0` (2018-09-18)

https://github.com/bitcrowd/sshkit.ex/compare/v0.0.3...v0.1.0

* Require Elixir 1.5+ and drop support for lower versions.
* Fix Elixir 1.5 String deprecations.
* Update installation instructions.
* Introduce Mox for mocking in unit tests.
* Improve and add unit tests.

Thanks for your contributions:

* @Bugagazavr
* @brienw
* @holetse

### Potentially breaking changes:

* Remove `:dry_run` option: Depending on how you're using SSHKit, "dry-run" could have a number of different meanings
  * you may want to actually connect to the remote without changing anything or you may not want to establish a connection at all
  * some steps in the flow you're dry-running may depend on things like directories created in a previous step which won't be there
  * all in all, a "dry-run" feature is likely better handled at an application level which may know the dependencies between commands
* Set `-H` option for `sudo` in order to get the expected value for `HOME`
* Export context environment variables directly before the supplied command in `SSHKit.Context.build/2`
  * this could potentially result in different behavior for code that sets environment variables consumed by the other commands in the context
  * those cases should be rare though and the new behavior seems closer to what most would expect when using contexts

### New features:

* Split the SCP file upload into a setup and an execution step:
  * add `SCP.Upload.new/3` and `SCP.Upload.exec/2`
  * `SCP.Upload.transfer/4` still works as before
* Added support for ptty allocation in `SSH.Channel` module [#129]

### Fixes:

* Fix error handling in `SSHKit.SSH.Channel.send/4` when sending stream data
* Context properly handles the case where env is set to an empty map
* Fix environment variables export in contexts with user, group, umask, path and env

## `0.0.3` (2017-07-13)

https://github.com/bitcrowd/sshkit.ex/compare/v0.0.2...v0.0.3

### New features:

* Added support for uploading/downloading files via the `SSHKIT` main DSL [#32]
* Added support for options that are shared between hosts [#61]
* Improved documentation [#67]
* Added support for passing an anonymous function to `SSH.connect` [#72]
* Add support for passing a `dry_run` flag to `SSHKit.SSH.connect` [#79]

### Fixes:

* Properly return `{:error, "No host given."}` when trying to connect to a host which is `nil` [#70]
* Improved unit and integration tests [#59] [#75] [#77]

## `0.0.2` (2017-06-23)

https://github.com/bitcrowd/sshkit.ex/compare/v0.0.1...v0.0.2

### Potentially breaking changes:

* Renamed response from remotely executed commands from 'normal' to 'stdout' [#34]
* Renamed `SSHKit.pwd` to `SSHKit.path` [#33] Thanks @brienw for the idea

### New features:

* Support basic SCP up-/downloads
* Added documentation https://hexdocs.pm/sshkit/SSHKit.html

### Fixes:

* Accept binaries (not only charlists) for configuration. Thanks @svoynow
* Fixed a bug that prevented `SSHKit.env` from working [#35]

## `0.0.1` (2017-01-26)

https://github.com/bitcrowd/sshkit.ex/releases/tag/v0.0.1

Basic support of / wrapping around erlang :ssh.


lib/sshkit.ex
defmodule SSHKit do
  @moduledoc """
  A toolkit for performing tasks on one or more servers.

  ```
  hosts = ["1.eg.io", {"2.eg.io", port: 2222}]

  context =
    SSHKit.context()
    |> SSHKit.path("/var/www/phx")
    |> SSHKit.user("deploy")
    |> SSHKit.group("deploy")
    |> SSHKit.umask("022")
    |> SSHKit.env(%{"NODE_ENV" => "production"})

  [:ok, :ok] = SSHKit.upload(context, ".", recursive: true)
  [{:ok, _, 0}, {:ok, _, 0}] = SSHKit.run(context, "yarn install", mode: :parallel)
  ```
  """

  alias SSHKit.Channel
  alias SSHKit.Connection
  alias SSHKit.Context
  alias SSHKit.Download
  alias SSHKit.Transfer
  alias SSHKit.Upload

  @doc """
  TODO

  Takes one or more (remote) hosts and creates an execution context in which
  remote commands can be run. Accepts any form of host specification also
  accepted by `host/1` and `host/2`, i.e. binaries, maps and 2-tuples.
  """
  @spec connect(binary(), keyword()) :: {:ok, Connection.t()} | {:error, term()}
  def connect(host, options \\ []) do
    Connection.open(host, options)
  end

  @spec close(Connection.t()) :: :ok
  def close(conn) do
    Connection.close(conn)
  end

  @spec exec!(Connection.t(), binary(), keyword()) :: Enumerable.t()
  def exec!(conn, command, options \\ []) do
    {context, options} = Keyword.pop(options, :context, Context.new())

    command = Context.build(context, command)

    # TODO: Separate options for open/exec/recv
    Stream.resource(
      fn ->
        # TODO: handle {:error, reason} and raise custom error struct?
        {:ok, chan} = Channel.open(conn, options)

        # TODO: timeout?, TODO: Handle :failure and {:error, reason} and raise custom error struct?
        :success = Channel.exec(chan, command)
        chan
      end,
      fn chan ->
        # TODO: timeout?, TODO: handle {:error, reason} and raise custom error struct?
        {:ok, msg} = Channel.recv(chan)

        # TODO: Adjust channel window size?

        value =
          case msg do
            {:exit_signal, ^chan, signal, message, lang} ->
              {:exit_signal, chan, signal, message, lang}

            {:exit_status, ^chan, status} ->
              {:exit_status, chan, status}

            {:data, ^chan, 0, data} ->
              {:stdout, chan, data}

            {:data, ^chan, 1, data} ->
              {:stderr, chan, data}

            {:eof, ^chan} ->
              {:eof, chan}

            {:closed, ^chan} ->
              {:closed, chan}
          end

        next =
          case value do
            {:closed, _} -> :halt
            _ -> [value]
          end

        {next, chan}
      end,
      fn chan ->
        :ok = Channel.close(chan)
        :ok = Channel.flush(chan)
      end
    )
  end

  # TODO: Do we need to expose lower-level channel operations here?
  #
  # * Send `eof`?
  # * Subsystem
  # * ppty
  # * …
  #
  # Seems like `send` and `eof` should be enough for the intended high-level use cases.
  # If more fine-grained control is needed, feel free to reach for the `SSHKit.Channel` module.

  @spec send(Channel.t(), :eof) :: :ok | {:error, term()}
  def send(chan, :eof) do
    Channel.eof(chan)
  end

  @spec send(Channel.t(), :stdout | :stderr, term(), timeout()) :: :ok | {:error, term()}
  def send(chan, type \\ :stdout, data, timeout \\ :infinity)
  def send(chan, :stdout, data, timeout), do: Channel.send(chan, 0, data, timeout)
  def send(chan, :stderr, data, timeout), do: Channel.send(chan, 1, data, timeout)

  @doc """
  TODO

  Accepts the same options as `exec!/3`.
  """
  @spec run!(Connection.t(), binary(), keyword()) :: [{:stdout | :stderr, binary()}]
  def run!(conn, command, options \\ []) do
    stream = exec!(conn, command, options)

    {status, output} =
      Enum.reduce(stream, {nil, []}, fn
        {:exit_status, _, status}, {_, output} -> {status, output}
        {:stdout, _, data}, {status, output} -> {status, [{:stdout, data} | output]}
        {:stderr, _, data}, {status, output} -> {status, [{:stderr, data} | output]}
        _, acc -> acc
      end)

    output = Enum.reverse(output)

    # TODO: Proper error struct?
    if status != 0, do: raise("Non-zero exit code: #{status}")

    output
  end

  @doc ~S"""
  Upload a file or files to the given context.

  Returns a list of `:ok` or `{:error, reason}` - one for each host.

  Possible options are:

  * `as: "remote.txt"` - specify the name of the uploaded file/directory
  * all options accepted by `SSHKit.SCP.Upload.transfer/4`

  ## Examples

  Upload all files and folders in current directory to "/workspace":

  ```
  [:ok] =
    "example.io"
    |> SSHKit.context()
    |> SSHKit.path("/workspace")
    |> SSHKit.upload(".", recursive: true)
  ```

  Upload a file to a different name on the remote host:

  ```
  [:ok] =
    "example.io"
    |> SSHKit.context()
    |> SSHKit.upload("local.txt", as: "remote.txt")
  ```
  """
  def upload!(conn, source, target, options \\ []) do
    Transfer.stream!(conn, Upload.init(source, target, options))
  end

  @doc ~S"""
  Download a file or files from the given context.

  Returns a list of `:ok` or `{:error, reason}` - one for each host.

  Possible options are:

  * `as: "local.txt"` - specify the name of the downloaded file/directory
  * all options accepted by `SSHKit.SCP.Download.transfer/4`

  ## Examples

  Download all files and folders in context directory to current working directory:

  ```
  [:ok] =
    "example.io"
    |> SSHKit.context()
    |> SSHKit.path("/workspace")
    |> SSHKit.download(".", recursive: true)
  ```

  Download a file to a different local name:

  ```
  [:ok] =
    "example.io"
    |> SSHKit.context()
    |> SSHKit.download("remote.txt", as: "local.txt")
  ```
  """
  def download!(conn, source, target, options \\ []) do
    # TODO
  end
end
